Added Max threshold to the default algorithm, and added a "do not engage" default algorithm

kkanzelmeyer opened the pull request:
I have a camera fixed outside watching an area of my house. It's been using this webcam-capture library for almost two years. Its been great! I noticed that when clouds pass in front of the sun it would cause the aperture of the webcam to open. Then when the clouds passed the camera would be flooded with light for one or two frames while the exposure adjusted to the light, which triggered a motion event. I was able to to mostly negate the problem by setting a fixed exposure value so that the exposure wouldn't change so much when the environmental light changes. However, it gave me the idea to implement a max threshold feature in the motion sensor algorithm so that sudden "white outs" caused by exposure would not trigger a motion event. The default max threshold is set to 100, effectively removing the limit. An accessor method has been added to allow the user to adjust the value.
![visitor20160816-133026](https://cloud.githubusercontent.com/assets/2906785/22312583/0b582dee-e31e-11e6-92d2-1e66d87d57b3.jpg)
![visitor20160816-131820](https://cloud.githubusercontent.com/assets/2906785/22312579/07382944-e31e-11e6-9e9a-fc2edd22ebbe.jpg)

Another challenge I encountered having a motion detection camera deployed is the hanging flower seen in the upper right of the image below. When the wind blows and the plant swings it would trigger a motion event (rightfully so). It gave me the idea to add a "Do Not Engage" feature to the algorithm. The "Do No Engage" feature is simply a rectangle specified by the user in which the algorithm will ignore motion. The feature is implemented by allowing the user to pass a `Rectangle` into the algorithm, and then the algorithm only processes pixels that are not contained in the `Rectangle`. 

![visitor20160601-154220](https://cloud.githubusercontent.com/assets/2906785/22312250/b8089efe-e31c-11e6-8f92-4e0e2e220f03.jpg)

sarxos commented on webcam-capture/src/main/java/com/github/sarxos/webcam/WebcamMotionDetector.java:
```diff
@@ -401,12 +401,14 @@ public void setDne(Rectangle rectangle) {
   {
-    if (detectorAlgorithm instanceof WebcamMotionDetectorDefaultAlgorithm
-        || detectorAlgorithm instanceof WebcamMotionDetectorDefaultWithDNE)
+    if (detectorAlgorithm instanceof WebcamMotionDetectorDefaultAlgorithm)
```
Hi @kkanzelmeyer,

This code is of course valid, but it would be better to have getThreasholds() extracted to an interface, such as:

```java
public interface ThresholdsAlgorithm {
  ArrayList<Integer> getThresholds();
}
```

And then, both WebcamMotionDetectorDefaultAlgorithm and WebcamMotionDetectorDefaultWithDNE can implement this interface. With this we can change this code to check instanceof interface instead of every class:

```java
public ArrayList<Integer> getThresholds()
  if (detectorAlgorithm instanceof ThresholdsAlgorithm) {
    return ((ThresholdsAlgorithm) detectorAlgorithm).getThresholds();
  }
  throw new UnsupportedOperationException("This method is only valid for algorithms which implements " + ThresholdsAlgorithm.class);
}
```

sarxos commented on webcam-capture-drivers/driver-gst1/pom.xml:
```diff
@@ -7,18 +7,19 @@
 			<name>sarxos-ftp-repo</name>
-			<url>http://sarxos.pl/maven2</url>
+			<!-- <url>http://repo.sarxos.pl/maven2</url> -->
+			<url>http://repo.sarxos.pl/maven2/index.php</url>
```
I think this should be without index.php. I'm not sure if this won't confuse Maven when it tries to download artifact from this repo.

sarxos commented:
Hi @kkanzelmeyer,

I downloaded your pull request, squashed multiplied commits into one, removed unnecessary changes and reworked it to remove other issues I found when testing. I also simplified the solution by removing  debugging mechanisms, integrating DNE feature into default motion detection algorithm and I also added example demonstrating how to use it.

You can find changes in:

https://github.com/sarxos/webcam-capture/commit/9ee9d94fde026832be4923e421a36ab90decd8b8

Since this feature is already merged into the master as one commit I will simply close this pull request.

Thank you Kevin!

kkanzelmeyer commented:
Very cool, thank you for adding the feature to the project!



Kevin Kanzelmeyer

On Fri, Aug 18, 2017 at 7:35 AM, Bartosz Firyn <notifications@github.com>
wrote:

> Hi @kkanzelmeyer <https://github.com/kkanzelmeyer>,
>
> I downloaded your pull request, squashed multiplied commits into one,
> removed unnecessary changes and reworked it to remove other issues I found
> when testing. I also simplified the solution by removing debugging
> mechanisms, integrating DNE feature into default motion detection algorithm
> and I also added example demonstrating how to use it.
>
> You can find changes in:
>
> 9ee9d94
> <https://github.com/sarxos/webcam-capture/commit/9ee9d94fde026832be4923e421a36ab90decd8b8>
>
> Since this feature is already merged into the master as one commit I will
> simply close this pull request.
>
> Thank you Kevin!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sarxos/webcam-capture/pull/524#issuecomment-323341898>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACxaoYuaSpS-0M7zKzaMqWpkF-Ai2XpZks5sZYUagaJpZM4LuGSA>
> .
>
